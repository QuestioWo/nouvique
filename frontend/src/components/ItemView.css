.item-image {
	width: 10rem;
}