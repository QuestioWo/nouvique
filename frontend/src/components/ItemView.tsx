import React from 'react';

import { Result } from 'neverthrow';
import { Button, Col, Row } from 'react-bootstrap';
import { RouteComponentProps } from 'react-router';

import { itemRESTLink, ItemREST, PageProps, resolveGETCall, itemTypeListRESTLink, ItemTypeREST, mediaURL } from '../utils';

import BasePage from './elements/BasePage';

import './ItemView.css';
import { FaShoppingBasket } from 'react-icons/fa';
import config from '../config';

interface MatchParams {
	name: string,
	username: string
}

export interface ItemViewProps extends RouteComponentProps<MatchParams>, PageProps { }

interface State extends ItemREST { }

export class ItemView extends React.Component<ItemViewProps, State> {
	constructor(props: ItemViewProps) {
		super(props);

		this.state = {
			name: "",
			bio: "",
			seller_name: "",
			upload_date: "",
			shape: "",
			colour: "",
			tag0: "",
			tag1: "",
			tag2: "",
			tag3: "",
			tag4: "",
			pictures: 0,
			item_types: []
		};
	}

	async componentDidMount() {
		const result: Result<ItemREST, Error> = await resolveGETCall<ItemREST>(itemRESTLink + this.props.match.params.username + "/" + this.props.match.params.name + "/");

		result
			.map(res => {
				this.setState({ ...res });

				return null; // necessary to silence warning
			})
			.mapErr(err => {
				console.error(err);
			});
	}

	handleAddToBasket = async () => {
		const resultType: Result<ItemTypeREST[], Error> = await resolveGETCall<ItemTypeREST[]>(itemTypeListRESTLink + this.props.match.params.username + "/" + this.props.match.params.name + "/");

		resultType
			.map(res => {
				var item_type = res[0];
				item_type.quantity = 1;
				this.props.addToBasket(this.state, item_type);

				return null; // necessary to silence warning
			})
			.mapErr(err => {
				console.error(err);
			});
	}

	render() {
		return (
			<React.Fragment>
				<BasePage {...this.props}>
					<Row>
						<h1 className="title">
							{this.props.match.params.name}
						</h1>
					</Row>

					<Row>
						<Col>
							<img alt="" className="item-image" src={config.apiURL + mediaURL + this.props.match.params.username + "/" + this.props.match.params.name + "/0.png"} />
						</Col>
						<Col>
							<Row>
								<Col>
									{this.state.bio}
									<br />
									{this.state.colour}
									<br />
									{this.state.shape}
									<br />
									{this.state.tag0}
									{' '}
									{this.state.tag1}
									{' '}
									{this.state.tag2}
									{' '}
									{this.state.tag3}
									{' '}
									{this.state.tag4}
								</Col>
								<Col>
									<div style={{ float: 'right' }}>
										<Button
											onClick={this.handleAddToBasket}
											disabled={this.props.match.params.username === localStorage.username}
											className="styled-btn" variant='outline-success'>
											<FaShoppingBasket /> add 1 to basket
									</Button>
									</div>
								</Col>
							</Row>
						</Col>
					</Row>
				</BasePage>
			</React.Fragment>
		);
	}
}