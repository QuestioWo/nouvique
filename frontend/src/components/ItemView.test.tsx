import * as React from 'react';

import { mount, shallow } from 'enzyme';

import { ItemView, ItemViewProps } from './ItemView';
import { MemoryRouter as Router } from 'react-router';
import { loadStripe } from '@stripe/stripe-js';

describe('ItemView', () => {
	const getProps = (): ItemViewProps => (
		{
			history: {} as any,
			location: {} as any,
			match: { params: { item_id: "foo" } as any } as any,
			updateAlertBar: async () => { },
			emptyBasket: () => { },
			addToBasket: () => { },
			removeFromBasket: () => { },
			getBasketItems: () => [],
			getTotalBasketPrice: () => 0,
			checkBasketAvailabilities: async () => { },
			stripePromise: loadStripe("foo"),
			alert: {} as any
		}
	)

	const getComponent = (props: ItemViewProps) => (
		<ItemView {...props} />
	);


	it('Should render correctly with shallow', () => {
		const component = shallow(
			getComponent(getProps())
		);
		expect(component.debug()).toMatchSnapshot();
	});

	it('Should render correctly with mount', () => {
		// have to include router as will fail with react-router problem due to use of specific functions
		const component = mount(
			<Router>
				{getComponent(getProps())}
			</Router>
		);
		expect(component.debug()).toMatchSnapshot();
	});
});