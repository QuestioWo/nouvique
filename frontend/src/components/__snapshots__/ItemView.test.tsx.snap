// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemView Should render correctly with mount 1`] = `
"<MemoryRouter>
  <Router history={{...}}>
    <ItemView history={{...}} location={{...}} match={{...}} updateAlertBar={[Function: updateAlertBar]} emptyBasket={[Function: emptyBasket]} addToBasket={[Function: addToBasket]} removeFromBasket={[Function: removeFromBasket]} getBasketItems={[Function: getBasketItems]} getTotalBasketPrice={[Function: getTotalBasketPrice]} checkBasketAvailabilities={[Function: checkBasketAvailabilities]} stripePromise={{...}} alert={{...}}>
      <BasePage history={{...}} location={{...}} match={{...}} updateAlertBar={[Function: updateAlertBar]} emptyBasket={[Function: emptyBasket]} addToBasket={[Function: addToBasket]} removeFromBasket={[Function: removeFromBasket]} getBasketItems={[Function: getBasketItems]} getTotalBasketPrice={[Function: getTotalBasketPrice]} checkBasketAvailabilities={[Function: checkBasketAvailabilities]} stripePromise={{...}} alert={{...}}>
        <div className=\\"master\\">
          <img alt=\\"\\" src=\\"/logo_main.png\\" className=\\"logo\\" />
           
          <NavBar>
            <Navbar bg=\\"transparent\\" variant=\\"light\\" expand={true} collapseOnSelect={false}>
              <nav className=\\"navbar navbar-expand navbar-light bg-transparent\\">
                <NavbarCollapse className=\\"justify-content-center\\">
                  <ForwardRef in={false} className=\\"justify-content-center\\" timeout={300} mountOnEnter={false} unmountOnExit={false} appear={false} getDimensionValue={[Function: getDefaultDimensionValue]}>
                    <ForwardRef addEndListener={[Function: transitionEndListener]} in={false} timeout={300} mountOnEnter={false} unmountOnExit={false} appear={false} aria-expanded={{...}} onEnter={[Function (anonymous)]} onEntering={[Function (anonymous)]} onEntered={[Function (anonymous)]} onExit={[Function (anonymous)]} onExiting={[Function (anonymous)]} childRef={{...}}>
                      <Transition in={false} timeout={300} mountOnEnter={false} unmountOnExit={false} appear={false} aria-expanded={{...}} onEnter={[Function (anonymous)]} onEntered={[Function (anonymous)]} onEntering={[Function (anonymous)]} onExit={[Function (anonymous)]} onExited={[Function (anonymous)]} onExiting={[Function (anonymous)]} addEndListener={[Function (anonymous)]} nodeRef={{...}} enter={true} exit={true}>
                        <div className=\\"justify-content-center navbar-collapse collapse\\" aria-expanded={{...}}>
                          <Nav className=\\"mr-auto nav_style\\" justify={false} fill={false}>
                            <ForwardRef as=\\"div\\" activeKey={[undefined]} className=\\"mr-auto nav_style navbar-nav\\" onSelect={[Function (anonymous)]}>
                              <div className=\\"mr-auto nav_style navbar-nav\\" onKeyDown={[Function: handleKeyDown]} role={[undefined]}>
                                <NavItem className=\\"item_style\\">
                                  <div className=\\"item_style nav-item\\">
                                    <NavLink className=\\"nav-link\\" isActive={[Function: isActive]} activeClassName=\\"nav-link active\\" to={{...}}>
                                      <Link aria-current=\\"page\\" className=\\"nav-link nav-link active\\" style={{...}} to={{...}}>
                                        <LinkAnchor aria-current=\\"page\\" className=\\"nav-link nav-link active\\" style={{...}} href=\\"/\\" navigate={[Function: navigate]}>
                                          <a aria-current=\\"page\\" className=\\"nav-link nav-link active\\" style={{...}} href=\\"/\\" onClick={[Function: onClick]}>
                                            Home
                                          </a>
                                        </LinkAnchor>
                                      </Link>
                                    </NavLink>
                                  </div>
                                </NavItem>
                                <NavItem className=\\"item_style\\">
                                  <div className=\\"item_style nav-item\\">
                                    <NavLink className=\\"nav-link\\" activeClassName=\\"nav-link active\\" to={{...}}>
                                      <Link aria-current={{...}} className=\\"nav-link\\" style={[undefined]} to={{...}}>
                                        <LinkAnchor aria-current={{...}} className=\\"nav-link\\" style={[undefined]} href=\\"/browse\\" navigate={[Function: navigate]}>
                                          <a aria-current={{...}} className=\\"nav-link\\" style={[undefined]} href=\\"/browse\\" onClick={[Function: onClick]}>
                                            Browse
                                          </a>
                                        </LinkAnchor>
                                      </Link>
                                    </NavLink>
                                  </div>
                                </NavItem>
                                <NavItem className=\\"item_style\\">
                                  <div className=\\"item_style nav-item\\">
                                    <NavLink className=\\"nav-link\\" activeClassName=\\"nav-link active\\" to={{...}}>
                                      <Link aria-current={{...}} className=\\"nav-link\\" style={[undefined]} to={{...}}>
                                        <LinkAnchor aria-current={{...}} className=\\"nav-link\\" style={[undefined]} href=\\"/ethics\\" navigate={[Function: navigate]}>
                                          <a aria-current={{...}} className=\\"nav-link\\" style={[undefined]} href=\\"/ethics\\" onClick={[Function: onClick]}>
                                            Ethics
                                          </a>
                                        </LinkAnchor>
                                      </Link>
                                    </NavLink>
                                  </div>
                                </NavItem>
                                <NavItem className=\\"item_style\\">
                                  <div className=\\"item_style nav-item\\">
                                    <NavLink className=\\"nav-link nav-right-side\\" activeClassName=\\"nav-link active nav-right-side\\" to={{...}}>
                                      <Link aria-current={{...}} className=\\"nav-link nav-right-side\\" style={[undefined]} to={{...}}>
                                        <LinkAnchor aria-current={{...}} className=\\"nav-link nav-right-side\\" style={[undefined]} href=\\"/login\\" navigate={[Function: navigate]}>
                                          <a aria-current={{...}} className=\\"nav-link nav-right-side\\" style={[undefined]} href=\\"/login\\" onClick={[Function: onClick]}>
                                            Login
                                          </a>
                                        </LinkAnchor>
                                      </Link>
                                    </NavLink>
                                  </div>
                                </NavItem>
                              </div>
                            </ForwardRef>
                          </Nav>
                        </div>
                      </Transition>
                    </ForwardRef>
                  </ForwardRef>
                </NavbarCollapse>
              </nav>
            </Navbar>
          </NavBar>
          <div className=\\"mainbody\\">
            <AlertBar history={{...}} location={{...}} match={{...}} updateAlertBar={[Function: updateAlertBar]} emptyBasket={[Function: emptyBasket]} addToBasket={[Function: addToBasket]} removeFromBasket={[Function: removeFromBasket]} getBasketItems={[Function: getBasketItems]} getTotalBasketPrice={[Function: getTotalBasketPrice]} checkBasketAvailabilities={[Function: checkBasketAvailabilities]} stripePromise={{...}} alert={{...}}>
              <Alert show={true} variant={[undefined]} onClose={[Function (anonymous)]} dismissible={true} transition={{...}} closeLabel=\\"Close alert\\">
                <Fade unmountOnExit={true} in={true} timeout={300} mountOnEnter={false} appear={false}>
                  <ForwardRef addEndListener={[Function: transitionEndListener]} unmountOnExit={true} in={true} timeout={300} mountOnEnter={false} appear={false} onEnter={[Function (anonymous)]} childRef={{...}}>
                    <Transition unmountOnExit={true} in={true} timeout={300} mountOnEnter={false} appear={false} onEnter={[Function (anonymous)]} onEntered={[Function (anonymous)]} onEntering={[Function (anonymous)]} onExit={[Function (anonymous)]} onExited={[Function (anonymous)]} onExiting={[Function (anonymous)]} addEndListener={[Function (anonymous)]} nodeRef={{...}} enter={true} exit={true}>
                      <div role=\\"alert\\" className=\\"fade alert alert-dismissible show\\">
                        <CloseButton onClick={[Function (anonymous)]} aria-label=\\"Close alert\\" variant={[undefined]}>
                          <button type=\\"button\\" className=\\"btn-close\\" onClick={[Function (anonymous)]} aria-label=\\"Close alert\\" />
                        </CloseButton>
                      </div>
                    </Transition>
                  </ForwardRef>
                </Fade>
              </Alert>
            </AlertBar>
            <Row>
              <div className=\\"row\\">
                <h1 className=\\"title\\" />
              </div>
            </Row>
            <Row>
              <div className=\\"row\\">
                <Col>
                  <div className=\\"col\\">
                    <img className=\\"item-image\\" src=\\"http://localhost:8000/api/media/undefined/undefined/0.png\\" />
                  </div>
                </Col>
                <Col>
                  <div className=\\"col\\">
                    <Row>
                      <div className=\\"row\\">
                        <Col>
                          <div className=\\"col\\">
                            <br />
                            <br />
                            <br />
                             
                             
                             
                             
                          </div>
                        </Col>
                        <Col>
                          <div className=\\"col\\">
                            <div style={{...}}>
                              <Button onClick={[Function (anonymous)]} disabled={true} className=\\"styled-btn\\" variant=\\"outline-success\\" active={false}>
                                <button onClick={[Function (anonymous)]} disabled={true} type=\\"button\\" className=\\"styled-btn btn btn-outline-success\\">
                                  <FaShoppingBasket>
                                    <IconBase attr={{...}}>
                                      <svg stroke=\\"currentColor\\" fill=\\"currentColor\\" strokeWidth=\\"0\\" viewBox=\\"0 0 576 512\\" className={[undefined]} style={{...}} height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\">
                                        <path d=\\"M576 216v16c0 13.255-10.745 24-24 24h-8l-26.113 182.788C514.509 462.435 494.257 480 470.37 480H105.63c-23.887 0-44.139-17.565-47.518-41.212L32 256h-8c-13.255 0-24-10.745-24-24v-16c0-13.255 10.745-24 24-24h67.341l106.78-146.821c10.395-14.292 30.407-17.453 44.701-7.058 14.293 10.395 17.453 30.408 7.058 44.701L170.477 192h235.046L326.12 82.821c-10.395-14.292-7.234-34.306 7.059-44.701 14.291-10.395 34.306-7.235 44.701 7.058L484.659 192H552c13.255 0 24 10.745 24 24zM312 392V280c0-13.255-10.745-24-24-24s-24 10.745-24 24v112c0 13.255 10.745 24 24 24s24-10.745 24-24zm112 0V280c0-13.255-10.745-24-24-24s-24 10.745-24 24v112c0 13.255 10.745 24 24 24s24-10.745 24-24zm-224 0V280c0-13.255-10.745-24-24-24s-24 10.745-24 24v112c0 13.255 10.745 24 24 24s24-10.745 24-24z\\" />
                                      </svg>
                                    </IconBase>
                                  </FaShoppingBasket>
                                   add 1 to basket
                                </button>
                              </Button>
                            </div>
                          </div>
                        </Col>
                      </div>
                    </Row>
                  </div>
                </Col>
              </div>
            </Row>
          </div>
        </div>
        <footer className=\\"footer py-3 text-black\\">
          <div />
          <Container fluid={false}>
            <div className=\\"container\\">
              <Row>
                <div className=\\"row\\">
                  <Col>
                    <div className=\\"col\\">
                      Nouvique 2021 v
                      0.22.0
                    </div>
                  </Col>
                  <Col>
                    <div className=\\"col\\">
                      <div style={{...}}>
                        <Link to=\\"/support\\">
                          <LinkAnchor href=\\"/support\\" navigate={[Function: navigate]}>
                            <a href=\\"/support\\" onClick={[Function: onClick]}>
                              Support
                            </a>
                          </LinkAnchor>
                        </Link>
                      </div>
                    </div>
                  </Col>
                </div>
              </Row>
            </div>
          </Container>
        </footer>
      </BasePage>
    </ItemView>
  </Router>
</MemoryRouter>"
`;

exports[`ItemView Should render correctly with shallow 1`] = `
"<Fragment>
  <BasePage history={{...}} location={{...}} match={{...}} updateAlertBar={[Function: updateAlertBar]} emptyBasket={[Function: emptyBasket]} addToBasket={[Function: addToBasket]} removeFromBasket={[Function: removeFromBasket]} getBasketItems={[Function: getBasketItems]} getTotalBasketPrice={[Function: getTotalBasketPrice]} checkBasketAvailabilities={[Function: checkBasketAvailabilities]} stripePromise={{...}} alert={{...}}>
    <Row>
      <h1 className=\\"title\\" />
    </Row>
    <Row>
      <Col>
        <img className=\\"item-image\\" src=\\"http://localhost:8000/api/media/undefined/undefined/0.png\\" />
      </Col>
      <Col>
        <Row>
          <Col>
            <br />
            <br />
            <br />
             
             
             
             
          </Col>
          <Col>
            <div style={{...}}>
              <Button onClick={[Function (anonymous)]} disabled={true} className=\\"styled-btn\\" variant=\\"outline-success\\" active={false}>
                <FaShoppingBasket />
                 add 1 to basket
              </Button>
            </div>
          </Col>
        </Row>
      </Col>
    </Row>
  </BasePage>
</Fragment>"
`;
